diff --git a/dist/typegraphql.module.js b/dist/typegraphql.module.js
index e09c77d1e8db56087886904fed4e037944ef37a9..bc687f64ead4478d2d9222a0f14edb359a9a4dc0 100644
--- a/dist/typegraphql.module.js
+++ b/dist/typegraphql.module.js
@@ -6,6 +6,7 @@ const tslib_1 = require("tslib");
 const common_1 = require("@nestjs/common");
 const graphql_1 = require("@nestjs/graphql");
 const constants_1 = require("./constants");
+const apollo = require('@nestjs/apollo')
 const prepare_options_service_1 = tslib_1.__importDefault(require("./prepare-options.service"));
 const typegraphql_options_factory_1 = tslib_1.__importDefault(require("./typegraphql-options.factory"));
 let TypeGraphQLModule = TypeGraphQLModule_1 = class TypeGraphQLModule {
@@ -20,6 +21,7 @@ let TypeGraphQLModule = TypeGraphQLModule_1 = class TypeGraphQLModule {
     }
     static forRoot(options = {}) {
         const dynamicGraphQLModule = graphql_1.GraphQLModule.forRootAsync({
+            driver: apollo.ApolloDriver,
             useClass: typegraphql_options_factory_1.default,
         });
         return Object.assign(Object.assign({}, dynamicGraphQLModule), { providers: [