// typegraphql-nestjs@0.4.1/dist/typegraphql-options-federation.factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const federation_1 = require("@apollo/federation");
const directives_1 = tslib_1.__importDefault(require("@apollo/federation/dist/directives"));
const common_1 = require("@nestjs/common");
const apollo_graphql_1 = require("apollo-graphql");
const graphql_1 = require("graphql");
const graphql_tag_1 = tslib_1.__importDefault(require("graphql-tag"));
const type_graphql_1 = require("type-graphql");
const prepare_options_service_1 = tslib_1.__importDefault(require("./prepare-options.service"));
const constants_1 = require("./constants");
let TypeGraphQLFederationOptionsFactory = class TypeGraphQLFederationOptionsFactory {
    constructor(rootModuleOptions, optionsPreparatorService) {
        this.rootModuleOptions = rootModuleOptions;
        this.optionsPreparatorService = optionsPreparatorService;
    }
    createGqlOptions() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { globalMiddlewares } = this.rootModuleOptions;
            const { resolversClasses, container, orphanedTypes, featureModuleOptionsArray, } = this.optionsPreparatorService.prepareOptions(constants_1.TYPEGRAPHQL_FEATURE_FEDERATION_MODULE_OPTIONS, globalMiddlewares);
            const referenceResolversArray = [...featureModuleOptionsArray].filter(it => it.referenceResolvers);
            const referenceResolvers = referenceResolversArray.length > 0
                ? Object.fromEntries(referenceResolversArray.flatMap(it => Object.entries(it.referenceResolvers)))
                : undefined;
            const baseSchema = yield type_graphql_1.buildSchema(Object.assign(Object.assign({}, this.rootModuleOptions), { directives: [...graphql_1.specifiedDirectives, ...directives_1.default], resolvers: resolversClasses, orphanedTypes,
                container }));
            const schema = federation_1.buildFederatedSchema({
                typeDefs: graphql_tag_1.default(federation_1.printSchema(baseSchema)),
                resolvers: type_graphql_1.createResolversMap(baseSchema),
            });
            if (referenceResolvers) {
                apollo_graphql_1.addResolversToSchema(schema, referenceResolvers);
            }
            return Object.assign(Object.assign({}, this.rootModuleOptions), { schema });
        });
    }
};
TypeGraphQLFederationOptionsFactory = tslib_1.__decorate([
    common_1.Injectable(),
    tslib_1.__param(0, common_1.Inject(constants_1.TYPEGRAPHQL_ROOT_FEDERATION_MODULE_OPTIONS)),
    tslib_1.__metadata("design:paramtypes", [Object, prepare_options_service_1.default])
], TypeGraphQLFederationOptionsFactory);
exports.default = TypeGraphQLFederationOptionsFactory;
